crates/pallet-transaction-fees/src/lib.rs

```diff
@@ -100,11 +112,19 @@ mod pallet {
     #[pallet::getter(fn storage_fees_escrow)]
     pub(super) type CollectedStorageFeesEscrow<T> = StorageValue<_, BalanceOf<T>, ValueQuery>;
 
-    /// Temporary value (cleared at block finalization) which contains cached value of
-    /// `TransactionByteFee` for current block.
+    /// Temporary value (it is set to `Some` during block execution and cleared at block finalization)
+    /// which contains cached value of `transaction_byte_fee` for current block.
     #[pallet::storage]
     pub(super) type TransactionByteFee<T> = StorageValue<_, BalanceOf<T>>;
 
+    /// The value of `transaction_byte_fee` for the next block (updated at block finalization),
+    /// it is used for validating extrinsic to be included in the next block, the value will move
+    /// into `TransactionByteFee` at block initialization to ensure using the exact same value for
+    /// both validation and execution.
+    #[pallet::storage]
+    pub(super) type NextTransactionByteFee<T> =
+        StorageValue<_, BalanceOf<T>, ValueQuery, InitialNextTransactionByteFee<T>>;
```

Comment:
This logic can be moved into `on_initialize` and that way `InitialNextTransactionByteFee` will not be necessary, I think it'll be a bit cleaner looking

Reply:
Not sure I understand, we need to update and return the value of `NextTransactionByteFee` after the execution of block #n (i.e. the block finalization) and before the execution of block #n+1 (i.e. the block initialization) for validating extrinsic to be included in block #n+1. And `InitialNextTransactionByteFee` is essentially for handling an edge case for the genesis block which doesn't have block finalization.

Reply:
That is correct. If you don't have `InitialNextTransactionByteFee` you can use `OptionValue` and use `.unwrap_or_else()` in initialization hook to do the same exact calculation as you do in `InitialNextTransactionByteFee`. Moreover you can add assertion that block number is `>1` because that is the only time when we don't expect the value to be present.

Reply:
>If you don't have InitialNextTransactionByteFee you can use OptionValue and use .unwrap_or_else() in initialization hook to do the same exact calculation

Actually, this is my first iteration of this PR, but then I realized when validating extrinsic to be included in block `#1` we are using the result of `calculate_transaction_byte_fee` and on the block initialization of `#1` we are calling the same `calculate_transaction_byte_fee` and use the result for executing extrinsic in block `#1` but the result of `calculate_transaction_byte_fee` may be different and bring the exact issue we try to fix in this PR for block `#1`.